### PR TITLE
Support inside properties for multi-instance completion condition

### LINKED ISSUE #2893: I can access the numberOf* multi-instance properties
**Is your feature request related to a problem? Please describe.**
The BPMN 2.0 specification defined the following properties of a multi-instance body instance:

* `numberOfInstances`
* `numberOfActiveInstances`
* `numberOfCompletedInstances`
* `numberOfTerminatedInstances`

**Describe the solution you'd like**
* I can access these properties inside the workflow instance / as a user

**Additional context**
* these properties are partially available in Camunda BPM 7: https://docs.camunda.org/manual/7.11/reference/bpmn20/tasks/task-markers/#multiple-instance

### PR DESCRIPTION
**Pull request by lzgabel**
## Description

The BPMN 2.0 specification defined the following properties of a multi-instance body instance:

* numberOfInstances
* numberOfActiveInstances
* numberOfCompletedInstances
* numberOfTerminatedInstances

This PR makes `numberOfInstances` and `numberOfActiveInstances` available for use in the [completion condition](https://docs.camunda.io/docs/components/modeler/bpmn/multi-instance/#completion-condition) expression. Although they are available in that expression, they do not exist as process variables. These properties take precedence over process variables with the same name.

Out of scope: `numberOfCompletedInstances`, `numberOfTerminatedInstances`
Out of scope: using these properties in other expressions

## Related issues

closes #2893

## Definition of Done

_Not all items need to be done depending on the issue and the pull request._

Code changes:
* [x] The changes are backwards compatibility with previous versions
* [ ] If it fixes a bug then PRs are created to [backport](https://github.com/zeebe-io/zeebe/compare/stable/0.24...develop?expand=1&template=backport_template.md&title=[Backport%200.24]) the fix to the last two minor versions. You can trigger a backport by assigning labels (e.g. `backport stable/0.25`) to the PR, in case that fails you need to create backports manually.

Testing:
* [x] There are unit/integration tests that verify all acceptance criterias of the issue
* [x] New tests are written to ensure backwards compatibility with further versions
* [x] The behavior is tested manually
* [x] The change has been verified by a QA run
* [ ] The impact of the changes is verified by a benchmark 

Documentation: 
* [ ] The documentation is updated (e.g. BPMN reference, configuration, examples, get-started guides, etc.)
* [x] New content is added to the [release announcement](https://drive.google.com/drive/u/0/folders/1DTIeswnEEq-NggJ25rm2BsDjcCQpDape)

### CODE REVIEW

**Comment by Lzgabel** [comment]
Hi @korthout , Can you take the time to review the code? :)

File: engine/src/main/java/io/camunda/zeebe/engine/processing/bpmn/container/MultiInstanceBodyProcessor.java
Comment (korthout) [inline engine/src/main/java/io/camunda/zeebe/engine/processing/bpmn/container/MultiInstanceBodyProcessor.java:511]: 🔧  I think we can merge this method with `wrapLoopCounter` into something like `wrapVariable` or so, because these do the same work

File: engine/src/main/java/io/camunda/zeebe/engine/processing/bpmn/container/MultiInstanceBodyProcessor.java
Comment (korthout) [inline engine/src/main/java/io/camunda/zeebe/engine/processing/bpmn/container/MultiInstanceBodyProcessor.java:552]: 🔧  We should reduce the number of read operations to the state as much as we can, because that can slow down things. We can do that by moving this call out of this method (i.e. it's already retrieved from state in `onChildActivating`, but not yet in `afterExecutionPathCompleted` and `onChildTerminated`). We can then have the value of `elementInstance.getNumberOfActiveElementInstances()` passed in as a parameter.

File: engine/src/main/java/io/camunda/zeebe/engine/processing/bpmn/container/MultiInstanceBodyProcessor.java
Comment (korthout) [inline engine/src/main/java/io/camunda/zeebe/engine/processing/bpmn/container/MultiInstanceBodyProcessor.java:518]: ❌  I don't think the `numberOfCompletedElementInstancesVariableView` is ever changed, so it doesn't keep track of this value yet. It is only initialized. We'll need some tests that show that the value is actually kept up to date with the reality.

To achieve it, we'll need to do something similar to what you already do in `updateNumberOfActiveElementInstances`, but then for the number of completed element instances. This is where things get more complicated.

The number of completed element instances is not tracked in `ElementInstance`, like the number of active children is. I see 2 solution directions:
1. read the current `numberOfCompletedElementInstances` local variable's value.
2. add the number of completed and terminated element instances as properties to `ElementInstance` and read them like you do the `numberOfActiveElementInstances` in `updateNumberOfActiveElementInstances`. 

I'd be curious if you see any other ways to solve this, but in any case I can already give you my thoughts on the solutions I suggest here. My initial feeling is that reading the local variable is bit error prone, because the user could alter this value leading to incorrect readings. 

Solution 2 sounds a bit better, because it means we get to keep track of the value in the state rather than through a user-modifiable variable, but there are other problems with it. Currently, `ElementInstance` only keeps track of the child count using `childCountProp`, we'll need to add something that represents the number of completed children. But the `ElementInstance` represents **any** element instance, not just multi-instance bodies. So we'd need to consider whether those new props should also be updated for **non-multi-instance** elements as well and  that might slow down processing. We'd need to verify this in some benchmarks, but let's postpone that for now and focus on getting it working first. We can optimize it later.

Lastly, we'll need to consider a migration path because existing Zeebe clusters will have multi-instance elements active that don't yet have these properties and did not yet keep track of it. This is true for both solution directions. Do we want to add them retroactively or simply ignore them for existing active multi-instances? I'd be happy to hear your thoughts on this.

**Review by korthout — CHANGES_REQUESTED** [review]
Hi @Lzgabel Thanks for giving this a go 🚀 . The first steps are already great, but there are some problems specifically with the `numberOfCompletedElementInstances` and `numberOfTerminatedElementInstances`, which will require some effort to resolve.

Therefore, I recommend to extract the changes for those properties into a separate PR again. The properties `numberOfElementInstances` and `numberOfActiveElementInstances` are much more straightforward and we might be able to get those merged much quicker.

The main thing that is missing for those 2 properties are some test cases. Please have a look at `MultiInstanceActivityTest.shouldSetLoopCounterVariable`. We'll need similar tests for the new properties.

I've also tried to explain the problems with the completed and terminated instances, to give you the chance to already think a bit about how to solve those problems. I think we'll need to do some back and forth discussing of solution ideas there before actually implementing it.

File: engine/src/main/java/io/camunda/zeebe/engine/processing/bpmn/container/MultiInstanceBodyProcessor.java
Comment (Lzgabel) [reply engine/src/main/java/io/camunda/zeebe/engine/processing/bpmn/container/MultiInstanceBodyProcessor.java:518]: Hi @korthout , As you mentioned, I think the second solution is more reasonable. Maybe we only need to add the number of completed element instances as properties to `ElementInstance`.

As we can see from page [#224](https://www.omg.org/spec/BPMN/2.0/PDF#page224), the numberOfTerminatedInstances variable of multi-instances are defined in the spec.
> The sum of `numberOfTerminatedInstances`, `numberOfCompletedInstances`, and `numberOfActiveInstances` always sums up to `numberOfInstances`.

So we can easily get the value of the `numberOfTerminatedInstances` element.

Comment (korthout) [reply engine/src/main/java/io/camunda/zeebe/engine/processing/bpmn/container/MultiInstanceBodyProcessor.java:518]: That's an interesting idea. 

The team has been increasingly more interested in adding data checks in the engine to make sure we don't end up in a corrupted state. This specific requirement may actually be an interesting property to check. Having all 4 variables available in state would allow us to add a consistency check for it. So that would be an argument for keeping track of both the `numberOfCompletedInstances` as well as the `numberOfTerminatedInstances `.

Comment (Lzgabel) [reply engine/src/main/java/io/camunda/zeebe/engine/processing/bpmn/container/MultiInstanceBodyProcessor.java]: Hi @korthout , Do I need to wait for the consistency check capability to be available before proceeding with follow-up work?

Comment (korthout) [reply engine/src/main/java/io/camunda/zeebe/engine/processing/bpmn/container/MultiInstanceBodyProcessor.java]: Hi @Lzgabel, I'm back from holidays 👋 

>Do I need to wait for the consistency check capability

No, I guess we had a miscommunication here. I mentioned this check as an argument to keeping track of both the `numberOfCompletedInstances` as well as the `numberOfTerminatedInstances` for each `ElementInstance`. 

However, adding these properties is complicated like I described. Therefore, I highly recommend to remove the `numberOfCompletedInstances` and the `numberOfTerminatedInstances` from this PR, and only focus on adding the `numberOfInstances` and the `numberOfActiveInstances` variables. That would allow us to quicker get new functionality merged as well as focus the other PR on the difficulties involved with the other variables. I hope that makes sense 🙇 

As always, please let me know if I can help with anything or if you have any questions. By the way, your contributions for completion condition will soon be available in release 1.3.0 🎉

**Comment by Lzgabel** [comment]
Hi @korthout , I've removed the `numberOfCompletedInstances` and the `numberOfTerminatedInstances` from this PR.   Please check this out. 🙇

File: engine/src/test/java/io/camunda/zeebe/engine/processing/bpmn/multiinstance/MultiInstanceActivityTest.java
Comment (korthout) [inline engine/src/test/java/io/camunda/zeebe/engine/processing/bpmn/multiinstance/MultiInstanceActivityTest.java:1363]: 👍 This test case is already good on it's own

❌ We should however consider how to test that the value is kept up to date with the actual number of active element instances. For example by adding an additional test, like this one, that checks the value after each completed job. 

🔧  When you've added a test case to verify that it should be updated, then you could consider changing the test you already wrote to verify its initial value (with Variableintent.CREATED instead of UPDATED).

🔧  Personally I appreciate group these tests near the `shouldSetLoopCounterVariable` testcase.

File: engine/src/main/java/io/camunda/zeebe/engine/processing/bpmn/container/MultiInstanceBodyProcessor.java
Comment (korthout) [inline engine/src/main/java/io/camunda/zeebe/engine/processing/bpmn/container/MultiInstanceBodyProcessor.java:43]: ❌  Let's use the same naming as defined in the spec: `numberOfInstances` and `numberOfActiveInstances`, respectively. These are also used by Camunda Platform 7.

**Review by korthout — CHANGES_REQUESTED** [review]
Thanks for the changes! Starting to look great @Lzgabel 🏗️ 

Please have a look at my comments. Like before, I'm using [the emoji code](https://github.com/camunda-cloud/zeebe/blob/develop/CONTRIBUTING.md#review-emoji-code) again. 

💭  I struggle with 1 part and that is the combination of completion condition and the `numberOfActiveInstances`. In the current form the `numberOfActiveInstances` variable is updated after the completion condition expression is evaluated. It seems better if we could do it the other way around. That would mean moving the updating of the `numberOfActiveInstances` variable to just before evaluation of the completion condition expression, but that leads to wrong changes if the instance doesn't fully complete (i.e. when an incident is raised in the meantime). I don't see ways to solve that problem yet, but I wanted to raise it. Perhaps you see any solutions here.

**Comment by Lzgabel** [comment]
Hi @korthout  , I've fixed all review comments, please check this out.  🙇

**Comment by korthout** [comment]
Thanks @Lzgabel. I'm currently out sick but I'll have a new look at it soon.

**Review by korthout — CHANGES_REQUESTED** [review]
Hi @Lzgabel 🎖️ I've finally had some time to review this and I'm happy to see you've found a solution and improved the tests.

However, in the case of an incident the `numberOfActiveInstances` variable is now updated twice (first lowered and then reset back to its previous value), which becomes confusing.

After a discussion with @saig0, we came to the conclusion that our completion logic does not currently allow us to use variables for these properties. Instead, we would like to propose to change the implementation to use "hidden" variables:

- currently, expressions are evaluated in the [ExpressionProcessor](https://github.com/camunda-cloud/zeebe/blob/develop/engine/src/main/java/io/camunda/zeebe/engine/processing/common/ExpressionProcessor.java#L346-L358)
- it uses an evaluation context ([VariableStateEvaluationContext](https://github.com/camunda-cloud/zeebe/blob/develop/engine/src/main/java/io/camunda/zeebe/engine/processing/common/ExpressionProcessor.java#L381-L407)), which in turn uses the variables stored in the state
- instead of storing these properties as variables in the state, we can try to find a way to provide additional variables to the VariableStateEvaluationContext
- these "hidden" variables should then be controlled by the MultiInstanceBodyProcessor: `numberOfInstances` and `numberOfActiveInstances`.
- this would mean these variables don't actually exist as variables in zeebe, but are available to expressions when they are evaluated. This way, the completion condition expression can still access these variables.

🙇 I understand that this requires quite a rewrite of the changes you've made here, but it would improve things. We actually think that we'll need to add more such "hidden" variables in the future, and these can be first ones. 🚀 

As always, please let me know if you have any questions or if I can help you in any way.

**Comment by korthout** [comment]
Hi @Lzgabel Can I help unblock you? Please let me know if there's anything I can do.

**Comment by Lzgabel** [comment]
Hi @korthout, Sorry for the delay caused by celebrating the Chinese New Year. I will continue this function next week. 🙇

Happy Lunar New Year!

**Comment by korthout** [comment]
@Lzgabel No worries whatsoever. Hope you enjoy the celebrations and happy lunar new year! 🎉

**Comment by Lzgabel** [comment]
Hi @korthout, I'm very sorry for the delay, please check this out. 🙇

File: engine/src/test/java/io/camunda/zeebe/engine/processing/bpmn/multiinstance/MultiInstanceActivityTest.java
Comment (korthout) [inline engine/src/test/java/io/camunda/zeebe/engine/processing/bpmn/multiinstance/MultiInstanceActivityTest.java:1017]: 🔧  The test verifies that `numberOfInstances` can be used in the `completionCondition` expression and that it has the correct value. I think your test case is fine, but it should be reflected in the test's method name.

❌ We also need a similar test for `numberOfActiveInstances`. It should make sure the value is correct, perhaps we should check it gets updated as well. I'm also curious what value it should have when the child instance completes. Will it be 0 or 1 in the case of a sequential multi-instance?

File: engine/src/main/java/io/camunda/zeebe/engine/processing/common/ExpressionProcessor.java
Comment (korthout) [inline engine/src/main/java/io/camunda/zeebe/engine/processing/common/ExpressionProcessor.java:122]: 🔧 Because you overload the `evaluateBooleanExpression` method, it makes sense to use this method as the implementation of the other `evaluateBooleanExpression(Expression, long)`, by passing in the `EMPTY_EVALUATION_CONTEXT`. This would remove a small bit of duplication.

**Review by korthout — CHANGES_REQUESTED** [review]
As always @Lzgabel ❤️  Thanks so much for this.

I'm really liking this approach for the multi-instance properties and I think it's almost ready. There's 1 larger problem still left to resolve and then we need an additional test case for `numberOfActiveInstances`. 

This will mark the start of an entirely new way of providing values to expressions (i.e. hidden variables). And once this is done, we should consider if there are other such variables we want to make available (e.g. `processInstanceKey` is a desired value to be used as the message event correlation key), and whether the multi-instance properties should also be available in input-output mappings of the multi-instance element. In any case, I think this is becoming an awesome feature! Thanks again for taking the time to build this 🚀

File: engine/src/test/java/io/camunda/zeebe/engine/processing/bpmn/multiinstance/MultiInstanceActivityTest.java
Comment (Lzgabel) [reply engine/src/test/java/io/camunda/zeebe/engine/processing/bpmn/multiinstance/MultiInstanceActivityTest.java:1017]: > Will it be 0 or 1 in the case of a sequential multi-instance?

Hi @korthout . Currently I don't find a definition for this situation in the bpmn specification, I wonder if we should be consistent with the `camunda-bpm-platform` behavior? From looking at the [technical documentation](https://docs.camunda.org/manual/latest/reference/bpmn20/tasks/task-markers/) and manual testing, the conclusion is that `numberOfActiveInstances` is always 1 when the child instance completes.

image

WDYT? Hope to give me a conclusion after internal discussion. 🙇

File: engine/src/main/java/io/camunda/zeebe/engine/processing/bpmn/container/MultiInstanceBodyProcessor.java
Comment (Lzgabel) [reply engine/src/main/java/io/camunda/zeebe/engine/processing/bpmn/container/MultiInstanceBodyProcessor.java:456]: I think the second option is probably the best option. 🤔

File: engine/src/test/java/io/camunda/zeebe/engine/processing/bpmn/multiinstance/MultiInstanceActivityTest.java
Comment (korthout) [reply engine/src/test/java/io/camunda/zeebe/engine/processing/bpmn/multiinstance/MultiInstanceActivityTest.java:1017]: @Lzgabel While we're discussing this topic internally, please just use the approach used in Camunda 7. However, note that I may come back at this with the request to change it.

File: engine/src/main/java/io/camunda/zeebe/engine/processing/bpmn/container/MultiInstanceBodyProcessor.java
Comment (korthout) [reply engine/src/main/java/io/camunda/zeebe/engine/processing/bpmn/container/MultiInstanceBodyProcessor.java:456]: Hi @Lzgabel I was rereading the spec on these properties again

> [about `numberOfInstances`] This attribute is provided for the outer instance of the Multi-Instance Activity only. This attribute contains the total number of inner instances created for the Multi-Instance Activity

So it is not the size of the input collection, but rather the number of created instances. 

This is further acknowledged by:

> [about `numberOfTerminatedInstances`] The sum of numberOfTerminatedInstances, numberOfCompletedInstances, and numberOfActiveInstances always sums up to numberOfInstances.

So in a sequential multi-instance with an input collection size of 3 and they all complete instead of terminate, the values count like this:
```
numberOfInstances 1, numberOfActiveInstances 1, numberOfCompletedInstances 0, numberOfTerminatedInstances 0
numberOfInstances 2, numberOfActiveInstances 1, numberOfCompletedInstances 1, numberOfTerminatedInstances 0
numberOfInstances 3, numberOfActiveInstances 1, numberOfCompletedInstances 2, numberOfTerminatedInstances 0
// and once the last one is completed
numberOfInstances 3, numberOfActiveInstances 0, numberOfCompletedInstances 3, numberOfTerminatedInstances 0
```

For a parallel multi-instance:
```
numberOfInstances 3, numberOfActiveInstances 3, numberOfCompletedInstances 0, numberOfTerminatedInstances 0
numberOfInstances 3, numberOfActiveInstances 2, numberOfCompletedInstances 1, numberOfTerminatedInstances 0
numberOfInstances 3, numberOfActiveInstances 1, numberOfCompletedInstances 2, numberOfTerminatedInstances 0
numberOfInstances 3, numberOfActiveInstances 0, numberOfCompletedInstances 3, numberOfTerminatedInstances 0
```

**Comment by Lzgabel** [comment]
Hi @korthout. I've fixed all comments, please check this out. 🙇

BTW. Seems we're having some problems with our CLA bot.

image

**Comment by Lzgabel** [comment]
Hi @korthout. I've fixed all comments, please check this out. 🙇

BTW. Seems we're having some problems with our CLA bot.

image

File: engine/src/main/java/io/camunda/zeebe/engine/state/instance/ElementInstance.java
Comment (korthout) [inline engine/src/main/java/io/camunda/zeebe/engine/state/instance/ElementInstance.java:25]: 🔧  I think `childActivatedCount` would be a better name in this context. Firstly, it could read as if the number represents all instances of this element instead of children of this specific element instance. And secondly, because it stands next to `childCount` and `childActivatedCount` would be somewhat consistent naming. If we expand it with `childCompletedCount` and `childTerminatedCount` in the future, it's somewhat easier to understand that `childCount` is the `childActiveCount`. 

To apply this further, we can already rename the field name from `childCountProp` to `childActiveCountProp`, but we cannot change the `childCount` name as that would be a breaking change.

**Review by korthout — CHANGES_REQUESTED** [review]
👍  Thanks @Lzgabel I like the new changes and we're getting closer.

❌  I have a single change request because existing zeebe clusters should be able to update to a version with these changes in it. So we'll need to find a solution to avoid the breaking change.

💭  This has made me think about supporting existing process instances. Next to the change I've requested, we'll probably need to implement a [MigrationTask](https://github.com/camunda-cloud/zeebe/blob/main/engine/src/main/java/io/camunda/zeebe/engine/state/migration/MigrationTask.java). The task should provide a value for the childActivatedCounter for the existing ElementInstances of type Multi-Instance. If we would not do this, then the value would be incorrect for those multi-instances. The implemented MigrationTask should then be added to the [DbMigrationImpl](https://github.com/camunda-cloud/zeebe/blob/main/engine/src/main/java/io/camunda/zeebe/engine/state/migration/DbMigratorImpl.java#L23-L28). Please let me know if you need help with that. If you feel comfortable with making those changes, then please do. Otherwise, I'd also be happy to contribute the implementation for this MigrationTask.

EDIT: please ignore the CLA assistant for now.

**Comment by korthout** [comment]
> BTW. Seems we're having some problems with our CLA bot.

@Lzgabel It seems our team has installed a new version that lost all previous signatures 😕 You'll have to sign it again. Sorry about that.

**Comment by Lzgabel** [comment]
Hi @korthout. I've fixed all comments, please check this out. 🙇

**Comment by korthout** [comment]
Hi @Lzgabel I'm sorry that I haven't been able to review this yet. We've been busy with the upcoming major release. I also have a few days off coming up, so I'll need to postpone this review to next week. I hope that's okay with you and again sorry for keeping you waiting.

File: protocol-impl/src/main/java/io/camunda/zeebe/protocol/impl/record/value/processinstance/ProcessInstanceRecord.java
Comment (korthout) [inline protocol-impl/src/main/java/io/camunda/zeebe/protocol/impl/record/value/processinstance/ProcessInstanceRecord.java:45]: ❌ This property has become unused, please remove it

File: engine/src/main/java/io/camunda/zeebe/engine/state/migration/DbMigrationState.java
Comment (korthout) [inline engine/src/main/java/io/camunda/zeebe/engine/state/migration/DbMigrationState.java:244]: 🔧 Let's add some documentation explaining how this works.

File: engine/src/main/java/io/camunda/zeebe/engine/state/migration/DbMigrationState.java
Comment (korthout) [inline engine/src/main/java/io/camunda/zeebe/engine/state/migration/DbMigrationState.java:243]: ❌ I don't think we should delete any of the entries of the parentChildColumnFamily. Updating the values like you do on lines 238 and 239 should be enough. Or am I missing something?

File: engine/src/main/java/io/camunda/zeebe/engine/state/migration/DbMigrationState.java
Comment (korthout) [inline engine/src/main/java/io/camunda/zeebe/engine/state/migration/DbMigrationState.java:240]: ❌ Here we'll also need to verify that the child has actually Activated. You can check this using the `getState()` method of the child ElementInstance.

File: engine/src/main/java/io/camunda/zeebe/engine/state/migration/MultiInstanceChildActivatedCountMigration.java
Comment (korthout) [inline engine/src/main/java/io/camunda/zeebe/engine/state/migration/MultiInstanceChildActivatedCountMigration.java:24]: Either the migration must be idempotent, or we should only run it exactly once. Even if it is idempotent, we should probably only run it if needed to save start-up time for existing clusters with large states.

❌ We should only run this migration once. As soon as it has run for all existing parents in the parentChildColumnFamily of type multi-instance it shouldn't be run again. So how do we check this?
I think the check should be something similar to the verification that's in the migration itself right now:

```java
parentChildColumnFamily.forEach(
        (key, value) -> {
          final long parentElementInstanceKey = key.first().getValue();
          if (parentElementInstanceKey != -1) {
            final var elementInstance = elementInstanceState.getInstance(parentElementInstanceKey);
            if (elementInstance != null
                && elementInstance
                    .getValue()
                    .getBpmnElementType()
                    .equals(BpmnElementType.MULTI_INSTANCE_BODY)) {
// now we know there exists an element instance of type multi-instance that has a child.
// we should also verify that a child has actually activated
// we should also verify that its `childActivatedCount` is still set to 0, which would only be the case for instances in the state that existed before these changes (0 is the default). 
```

**Review by korthout — CHANGES_REQUESTED** [review]
Hi @Lzgabel I finally got around to reviewing this. Please accept my apology for the delays 🙇 After my previous message I was on sick leave for some time. But I'm happy to be back 🎉 

👏 Thanks again for these changes. It's really looking good and you've done a great job with the migration code.

❌ I've highlighted a few things that I encountered that we should fix before merging. Please have a look at them. 

🔧 I also want to propose that you rebase your PR on top of #9118 (or wait until it is merged). This PR adds a similar construct as you've introduced that allows adding additional variable lookups to the expression processor. I would like to see that this PR uses it.

❓ Lastly, I imagine that it must be frustrating that this PR has been open for so long. I'm very happy to receive your contribution and I want to encourage you to continue. But I would also understand it if you'd like me to take this PR over from you. Once it's merged, we could focus on the next steps like the `numberOfCompletedInstances` and `numberOfTerminatedInstances`. Please let me know, whether you want to continue with it yourself or would prefer me to take it over. 

As always thanks a lot for your efforts. They're greatly appreciated 🙇

**Comment by Lzgabel** [comment]
Hi @korthout. Thanks for your reply.  I want to continue with it.🧑🏻‍💻

**Comment by Lzgabel** [comment]
Hi @korthout. Sorry for the long wait, i've rebased this changes on top of `main`,  it may require you to re-review the code, sorry for the inconvenience, please check this out. 🙇

File: engine/src/main/java/io/camunda/zeebe/engine/processing/bpmn/container/MultiInstanceBodyProcessor.java
Comment (korthout) [inline engine/src/main/java/io/camunda/zeebe/engine/processing/bpmn/container/MultiInstanceBodyProcessor.java:389]: 🔧 Please add a comment on why this variable has to be lowered by 1. I think it's because the `getNumberOfActiveInstancesVariable` method is called while the child instance is completing but the active element instances value has not yet been decremented.

File: engine/src/test/java/io/camunda/zeebe/engine/processing/bpmn/multiinstance/MultiInstanceActivityTest.java
Comment (korthout) [inline engine/src/test/java/io/camunda/zeebe/engine/processing/bpmn/multiinstance/MultiInstanceActivityTest.java:1085]: 🔧 I believe that this test case for the parallel loopCharacteristics actually works best when you complete 2 of the 3 jobs. Once the 2nd instance completes, the completionCondition evaluates to true and then terminates the last instance. The reason that the test also works with 3 is that the engine seems to run faster than the 3rd complete job command can be sent and processed.

**Review by korthout — CHANGES_REQUESTED** [review]
🚀 🚀 🚀  Great work @Lzgabel! This looks awesome! 

I have one final request and then we can merge. Last round, I promise 🙈 Please also consider my suggestions.

**Review by korthout — APPROVED** [review]
LGTM 🚀 

Thanks for all your efforts! We really appreciate it ❤️ 

PS: I've updated the PR description and title to reflect the current state

**Comment by korthout** [comment]
@Lzgabel there appears to be a formatting error, please run `mvn spotless:apply` and just squash it.

**Comment by korthout** [comment]
bors merge

**Pull request merged**
